src/nvim/eval/typval.c

```diff
@@ -1208,6 +1213,9 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       watcher->busy = true;
       callback_call(&watcher->callback, 3, argv, &rettv);
       watcher->busy = false;
+      if (watcher->needs_free) {
+        tv_dict_watcher_free(watcher);
+      }
```

Comment:
Does this mean `dictwatcherdel` would need to wait for the notification to trigger again, before the object is actually freed? I guess that could be surprising for someone watching memory usage after calling `dictwatcherdel`.

Reply:
I don't see the problem, is this not synchronous already? To me it looks like this free call will free the dict watcher in the _same_ event as the `callback_call` which caused it.

Reply:
Yes.